### DISCUSSION TITLE
S0/S2 security - pros and cons

### DISCUSSION THREAD

**Discussion by ivdimitro** [zwave-js/zwave-js#3729] [Q&A] [answered]
Hello, I really appreciate the work that has been done to support the S0/S2 security in zwaveJS but I am wondering what are the practical implications of actually using it? 
Personally I have a few TRVs and a few sensors so I am not bothered if somebody snoop my trafic. On the other hand I see some theoretical concerns and practical questions: 

1. implementing any kind of encryption requires more processing power --> reduced battery life of my devices. Is this correct? 
2. I have a mix of devices supporting S0 secure communication and S2 - can the controller handle it gracefully? 
3. All of my devices were included in the network before the security was implemented and I couldn't find "migration guide". I guess I will have to recreate my network from scratch in order to securely include all devices.  

Oveal I don't see much benefit of the security in my case. Am I missing something? Are there other benefits of which I am not aware? 
Thanks. 

P.S. Don't get me wrong - I am NOT saying security is pointless. Just in my case I can't justify the extra work of recreating the network from scratch.

**Comment by AlCalzone** [comment, marked as answer]
1.: I would assume that it adds very little overhead, if any. Z-Wave nodes do encryption in hardware, so this is very performant.

2.: The controller communcation is well-tested, but there are some pitfalls when it comes to direct associations. While the specs mention that a node controlling another one with a lower security class (i.e. S2 -> S0) should detect which security class the target node has, this is just a **SHOULD** and some devices don't do this. Also direct associations to control a node with a higher security class won't work (S0 -> S2).

3.: The recommended approach is to shutdown the host (to prevent device reset locally notifications from removing the node ID), reset each device, then re-include with S2 using the replace failed node function. There appears to be another way by using a secondary controller (included in the same network) with a pushbutton (like Aeotec Gen5). Apparently that one can be used to exclude a node without the primary controller knowing. You can then re-include with S2 using the replace failed node function.

---

Over all, S2 adds some key benefits with very little overhead, which is why we recommend it by default (unless S0).
- Encryption obviously
- Inclusion via QR code / Smart Start is super painless
- The communication is much more performant than S0 (no nonce exchange after the node and controller are in sync)
- Encryption includes integrity checks by default, avoiding measurements from sensors that got corrupted on air
- The controller will know when a destination wasn't able to decrypt a command (unlike S0)
- S2 capable nodes support Supervision (sometimes only in S2 mode), allowing the driver to reduce traffic by avoiding verification GETs

---

Its up to you if you want to do the work. If you were starting anew, I'd say go for S2.

**Reply by ivdimitro** [reply]
Thanks a lot for the clarification. I wasn't aware of the communication benefits. The holidays are coming so I might give the S2 a try.

**Comment by baylanger** [comment]
In case others stumble on this discussion, I found [this note](https://zwave-js.github.io/node-zwave-js/#/troubleshooting/network-health?id=optimizing-the-reporting-configuration) from the docs.

[edit] See note below from @AlCalzone 

> Do not use Security S0 for sensors. S0 is very inefficient and devices using S0 need to exchange three commands for each report. If encryption is critical, use devices that support S2, otherwise stick to not using encryption. The only exception is if your device frequently reports nonsensical values (like water usage on a power meter), which can happen when commands get corrupted on air. S0 (and S2) prevents this. If possible, configure your device to use CRC-16 encapsulation instead, which doesn't require additional messages.

**Reply by AlCalzone** [reply]
I should reword this part a bit. We do recommend using S2, since it provides encryption and data integrity. This is reflected in the default inclusion strategy.

**Reply by AlCalzone** [reply]
https://github.com/zwave-js/node-zwave-js/pull/6770
